### PR TITLE
Feature/ntgl 820 friend group

### PR DESCRIPTION
**Pull request by jasttran**
Friend Features:
- see friends list and can remove friends
- see incoming friend requests
- search for users and send friend requests

Group Features:
- can create a group with friends.
- admins can edit/delete groups
- members can leave groups

### CODE REVIEW

File: client/src/App.tsx [hunk 1]
```diff
@@ -266,9 +266,9 @@ const App: React.FC = () => {
       Object.keys(course.activities).forEach((activity) => {
```
Comment (Rayahhhmed) [inline client/src/App.tsx]: break this code into functions, very hard to read oops

File: client/src/components/sidebar/Sidebar.tsx [hunk 2]
```diff
@@ -88,9 +100,29 @@ const SidebarFooterText = styled('div')`
...
+    display: none; /* Chrome, Safari, and Edge */
+  }
+`;
+
 const Sidebar: React.FC = () => {
   const [currLogo, setCurrLogo] = useState(notanglesLogo);
   const [collapsed, setCollapsed] = useState(true);
+  const [showGroupsSidebar, setShowGroupsSidebar] = useState(false);
   const sideBarRef = useRef<HTMLDivElement>(null);
   // TODO: dummy logic to be
```
Comment (Rayahhhmed) [inline client/src/components/sidebar/Sidebar.tsx]: wats this TODO? Could you remove it :D

File: client/src/components/sidebar/groupsSidebar/AddOrEditGroupDialogContent.tsx [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+  user,
+  group,
+  setGroup,
+  handleClose,
+  isEditing,
+}) => {
+  const [errorMessage, setErrorMessage] = useState('');
+
+  const checkInputs = () => {
+    console.log('groupp', group);
```
Comment (Rayahhhmed) [inline client/src/components/sidebar/groupsSidebar/AddOrEditGroupDialogContent.tsx]: remove clg

File: client/src/components/sidebar/groupsSidebar/AddOrEditGroupDialogContent.tsx [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    if (group.name.length < 1) {
+      setErrorMessage(InputError.GROUP_NAME);
```
Comment (Rayahhhmed) [inline client/src/components/sidebar/groupsSidebar/AddOrEditGroupDialogContent.tsx:87]: rlly good code here with the enums

File: client/src/components/sidebar/groupsSidebar/friends/YourFriendsTab.tsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  flex-direction: column;
+  gap: 12px;
+`;
+
+const StyledItem = styled('div')`
+  display: flex;
+  justify-content: space-between;
+`;
+
+const YourFriendsTab: React.FC<{ user: User | undefined; getUserInfo: () => void }> = ({ user, getUserInfo }) => {
```
Comment (Rayahhhmed) [inline client/src/components/sidebar/groupsSidebar/friends/YourFriendsTab.tsx:22]: sadly dont have any

File: client/src/components/sidebar/groupsSidebar/friends/YourFriendsTab.tsx [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        headers: {
+          Accept: 'application/json',
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify({
+          senderId: user.userID,
+          sendeeId: friendID,
+        }),
+      });
+      //   if (res.status !== 200) throw new NetworkError("Couldn't get response");
```
Comment (Rayahhhmed) [inline client/src/components/sidebar/groupsSidebar/friends/YourFriendsTab.tsx]: is this comment supposed to be here

File: client/src/components/sidebar/groupsSidebar/friends/YourFriendsTab.tsx [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+const StyledItem = styled('div')`
+  display: flex;
+  justify-content: space-between;
+`;
+
+const YourFriendsTab: React.FC<{ user: User | undefined; getUserInfo: () => void }> = ({ user, getUserInfo }) => {
+  if (!user) return <></>;
+
+  const handleRemoveFriend = async (friendID: string) => {
+    try {
```
Comment (Rayahhhmed) [inline client/src/components/sidebar/groupsSidebar/friends/YourFriendsTab.tsx:26]: would a function like 3900 help here?

File: server/src/user/user.service.ts [hunk 8]
```diff
@@ -320,4 +322,60 @@ export class UserService {
...
+          },
+        });
+
+        // convert members, groupAdmins and timetables objects into string ids
+        group.memberIDs = group.members.map((group) => group.userID);
+        group.groupAdminIDs = group.groupAdmins.map(
+          (groupAdmin) => groupAdmin.userID,
+        );
+        group.timetableIDs = group.timetables.map((timetable) => timetable.id);
+        delete group.members;
```
Comment (Rayahhhmed) [inline server/src/user/user.service.ts:357]: `delete` is a scary keyword, what is this supposed to do?

**Review by Rayahhhmed — CHANGES_REQUESTED** [review]

File: client/src/components/sidebar/Sidebar.tsx (hunk 2 above)
Comment (jasttran) [reply client/src/components/sidebar/Sidebar.tsx]: not sure it was there before, but i shall remove its existence 🫡

File: client/src/components/sidebar/groupsSidebar/friends/YourFriendsTab.tsx (hunk 6 above)
Comment (jasttran) [reply client/src/components/sidebar/groupsSidebar/friends/YourFriendsTab.tsx]: yeah thanks i uncommented it

File: server/src/user/user.service.ts (hunk 8 above)
Comment (jasttran) [reply server/src/user/user.service.ts:357]: the `delete` keyword is used to remove properties from an object. This is a JavaScript feature that allows you to modify an object by removing specific keys.

This part of the code is removing the members, groupAdmins, and timetables properties from the group object.

This gets removed in ticket #880 though.

File: client/src/components/sidebar/groupsSidebar/friends/YourFriendsTab.tsx (hunk 7 above)
Comment (jasttran) [reply client/src/components/sidebar/groupsSidebar/friends/YourFriendsTab.tsx:26]: yeah it would

File: client/src/App.tsx (hunk 1 above)
Comment (jasttran) [reply client/src/App.tsx]: ermm I might need help on this, cuz I don't really know what you mean

**Review by Rayahhhmed — APPROVED** [review]

**Review by Shzmj — APPROVED** [review]
AMAZING WORK 🚀. 
Nit: console.log statements within a few files (AddAFriendTab, RequestsTab) for response status -> maybe we would need these when we do final testing for friends system? @Rayahhhmed

**Comment by Rayahhhmed** [comment]
@jasttran  run a final lint as well before u push
We can have a logger installed to mitigate this @Shzmj  so during prod we don't see dev console logs. I will look into it!

**Pull request merged**